### PR TITLE
[Service Bus] Allow 0 prefetch and dynamically use batch size to request link credits

### PR DESCRIPTION
**Pull request by YijunXieMS**
This PR is for sync receiving performance tuning and allow prefetch 0 to avoid the bad side effects of using a non-zero prefetch.

1. 0 prefetch is allowed, as in other languages. For ServiceBus, using prefetch risks of missing some messages especially in the RECEIVE_AND_DELETE mode. This is different from EventHubs, with which one can read an event for unlimited times.
2. When the prefetch is 0, this sync API add link credits by maxMessages to remarkably improve performance as opposed to adding link credit one by one.
```
public IterableStream<ServiceBusReceivedMessage> receiveMessages(int maxMessages, Duration maxWaitTime)
```
3. Fixes a bug that adds too many link credits.

### CODE REVIEW

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java [hunk 1]
```diff
@@ -590,13 +590,14 @@ public void close() {
     private void queueWork(int maximumMessageCount, Duration maxWaitTime,
                            FluxSink<ServiceBusReceivedMessage> emitter) {
         final long id = idGenerator.getAndIncrement();
-        final SynchronousReceiveWork work = new SynchronousReceiveWork(id, maximumMessageCount, maxWaitTime, emitter);
-
+        int prefetch = asyncClient.getReceiverOptions().getPrefetchCount();
```
Comment (hemanttanwar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java:593]: final ?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java [hunk 2]
```diff
@@ -199,12 +198,7 @@ public void onNext(ServiceBusReceiveLink next) {
             oldSubscription = currentLinkSubscriptions;
 
             currentLink = next;
-            next.setEmptyCreditListener(() -> {
-                final int creditsToAdd = getCreditsToAdd(0);
-                linkCreditsAdded.set(creditsToAdd > 0);
-
-                return creditsToAdd;
-            });
+            next.setEmptyCreditListener(() -> 0);
```
Comment (hemanttanwar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:201]: Should we do similar logic in SessionManager, Like we spoke over the chat.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 3]
```diff
@@ -107,6 +107,11 @@ public void addCredits(int credits) {
         }
     }
 
+    @Override
+    public void addCreditsInstantly(int credits) {
+        receiver.flow(credits);
```
Comment (hemanttanwar) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:112]: Most of the calls into Proton-j Reactor is done via ReactorDispatcher because of thread safety,  
As commented here ..
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorDispatcher.java#L22
Would this be okay to call this api directly ? @srnagar  Would this be okay here ?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java (hunk 2 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:202]: Updated to 0 in ServiceBusSessionReceiver

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java (hunk 1 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java]: Updated to final

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionReceiver.java [hunk 4]
```diff
@@ -63,14 +63,23 @@
...
-                receiveLink.addCredits(prefetch);
+                if (prefetch > 0) {
+                    receiveLink.addCreditsInstantly(prefetch);
+                }
+            })
+            .doOnRequest(request -> {  // request is of type long.
+                if (prefetch == 0) {  //  add "request" number of credits
+                    receiveLink.addCreditsInstantly((int) request);
+                } else {  // keep total credits "prefetch" if prefetch is not 0.
+                    receiveLink.addCreditsInstantly(Math.max(0, prefetch - receiveLink.getCredits()));
```
Comment (hemanttanwar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionReceiver.java:81]: Just like we have `addCreditsInstantly` , adding locally , Does `getCredits` return from local? And if it is returning from remote, there could be short time when there is diff in local v/s remote .
Should there is equivalent `getCreditsInstantly` ?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java [hunk 5]
```diff
@@ -571,21 +562,28 @@ private int getCreditsToAdd(int linkCredits) {
...
 
         final int creditsToAdd;
-        if (messageQueue.isEmpty() && !hasBackpressure) {
-            creditsToAdd = prefetch;
+        final int expectedTotalCredit;
+        if (prefetch == 0) {
+            if (r <= Integer.MAX_VALUE) {
+                expectedTotalCredit = (int) r;
+            } else {
+                expectedTotalCredit = Integer.MAX_VALUE;
```
Comment (hemanttanwar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:570]: How we get this in this case where  credit will need to be `Integer.MAX_VALUE` ? And should it be max ?

**Review by hemanttanwar — COMMENTED** [review]
Can you test live test on your local for `receiverIntegrationTest`. To ensure we did not have any regression.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionReceiver.java (hunk 4 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionReceiver.java:81]: getCredits() should return from local. I don't see a log for traffic of getCredits.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java (hunk 5 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:570]: This won't really happen in reality. I still leave it there because Integer.MAX_VALUE is the largest allowed link credits.
For async client, receiveMessages() calls "return receiveMessagesNoBackPressure().limitRate(1, 0);". This will request one by one from the link processor, even though the user's request has no back pressure.

For sync client, the sync subscriber has back pressure. The request count uses the  the argument of method receiveMessages(int maxMessages). It's at most Integer.MAX_VALUE.

Comment (hemanttanwar) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:570]: So does this justify In that case should we log point to help us debug this if needed.

Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java:577]: Added the code comments.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusAsyncConsumer.java [hunk 6]
```diff
@@ -34,9 +34,7 @@
         this.linkProcessor = linkProcessor;
         this.messageSerializer = messageSerializer;
         this.processor = linkProcessor
-            .map(message -> this.messageSerializer.deserialize(message, ServiceBusReceivedMessage.class))
-            .publish(receiverOptions.getPrefetchCount())
-            .autoConnect(1);
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusAsyncConsumer.java:39]: Why was this removed?

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java [hunk 7]
```diff
@@ -849,12 +849,15 @@ void canPerformMultipleReceive() {
...
-            .then(() -> messages.forEach(m -> messageSink.next(m)))
-            .expectNextCount(numberOfEvents)
-            .verifyComplete();
+        // TODO: Do we need to support multiple calls of receiver.receiveMessages()?
+        //  After the autoConnect was removed from ServiceBusAsyncConsumer.processor, the receiver doesn't support
+        //  multiple calls of receiver.receiveMessages().
+//        StepVerifier.create(receiver.receiveMessages().take(numberOfEvents))
+//            .then(() -> messages.forEach(m -> messageSink.next(m)))
+//            .expectNextCount(numberOfEvents)
+//            .verifyComplete();
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java:858]: We need to support multiple calls to `receiveMessages`. So, this should be fixed.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java [hunk 8]
```diff
@@ -150,8 +150,8 @@ private synchronized void receiveMessages() {
         }
         ServiceBusReceiverAsyncClient receiverClient = asyncClient.get();
         receiverClient.receiveMessagesWithContext()
-            .parallel(processorOptions.getMaxConcurrentCalls())
-            .runOn(Schedulers.boundedElastic())
+            .parallel(processorOptions.getMaxConcurrentCalls(), 1)
+            .runOn(Schedulers.boundedElastic(), 1)
```
[the remark this answers is not in the document]
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java]: Updated and pushed after discussion

File: eng/versioning/version_client.txt [hunk 9]
```diff
@@ -177,7 +177,8 @@ com.microsoft:microsoft-opentelemetry-exporter-azuremonitor;1.0.0-beta.1;1.0.0-b
 # unreleased_<groupId>:<artifactId>;dependency-version
 # note: The unreleased dependencies will not be manipulated with the automatic PR creation code.
 unreleased_com.azure:azure-core-experimental;1.0.0-beta.9
-unreleased_com.azure:azure-messaging-servicebus;7.0.0-beta.7
+unreleased_com.azure:azure-messaging-servicebus;7.0.0-beta.8
```
[the remark this answers is not in the document]
Comment (YijunXieMS) [reply eng/versioning/version_client.txt:180]: 7.0.0-beta.7 has been released so I updated it when I updated the core-amqp unreleased. This PR shouldn't require it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusAsyncConsumer.java (hunk 6 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusAsyncConsumer.java:39]: This is removed because the publish( ) method requests the upstream (link processor) with its own request count (the reactor prefetch). Without publish, if a user's code subscriber request n records, the link processor can adjust the link credits with by the user's request count. So with this publish(), we won't be able to adjust the link credits by the user's request count.
Plus, receiverOptions.getPrefetchCount() can be 0 now and defaults to be 0. It's an invalid value for publish() method.

The side effect of no-publish is a receiveMessages can't be published twice. The link processor will throw an error if there are two subscribers. 
We need to discuss offline.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java (hunk 7 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java]: Fixed after I add publish().autoConnect() to receiveMessages() implementation.

File: eng/versioning/version_client.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (YijunXieMS) [reply eng/versioning/version_client.txt]: Removed it after merging from master.

**Review by hemanttanwar — APPROVED** [review]
LGTM

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java (hunk 3 above)
Comment (srnagar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:112]: This needs to be tested for thread-safety. If proton-j doesn't support adding credits in a thread-safe manner we might add incorrect number of credits to the link and can potentially result in data loss if the SB mode is RECEIVE_AND_DELETE

**Review by srnagar — COMMENTED** [review]
I think there is one concern about thread-safety with proton-j. If you have verified this and if it's not an issue, then we can go ahead with the PR.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java (hunk 3 above)
Comment (YijunXieMS) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:112]: I simply put it into a synchronized block. The performance overhead should be minimal.

Comment (YijunXieMS) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:112]: Changed back to addCredits()

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java (hunk 7 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java]: Discussed with Srikanta again. We agreed to remove the publish and autoConnect and don't support multiple subscribers now.
In the future, we may create a link for each subscriber. To be discussed in the future.
